tutorials/tour/_posts/2017-02-13-mixin-class-composition.md

```diff
@@ -39,17 +52,28 @@ class StringIterator(s: String) extends AbsIterator {
   def next = { val ch = s charAt i; i += 1; ch }
 }
 ```
- 
-We would like to combine the functionality of `StringIterator` and `RichIterator` into a single class. With single inheritance and interfaces alone this is impossible, as both classes contain member implementations with code. Scala comes to help with its _mixin-class composition_. It allows the programmers to reuse the delta of a class definition, i.e., all new definitions that are not inherited. This mechanism makes it possible to combine `StringIterator` with `RichIterator`, as is done in the following test program which prints a column of all the characters of a given string.
- 
+`StringIterator` takes a `String` and can be used to iterate over a String (e.g. to see if a String contains a certain character). `type T` must be `Char` so we know the return type of `next`.
+
+Now let's create a mixin class which also extends `AbsIterator`.
+
+```tut
+trait RichIterator extends AbsIterator {
+  def foreach(f: T => Unit) { while (hasNext) f(next) }
+}
+```
+Because `RichIterator` is a trait, it doesn't need to implement the abstract members of AbsIterator.
+
+We would like to combine the functionality of `StringIterator` and `RichIterator` into a single class.
+
 ```tut
 object StringIteratorTest {
   def main(args: Array[String]) {
-    class Iter extends StringIterator(args(0)) with RichIterator
-    val iter = new Iter
+    class RichStringIter extends StringIterator(args(0)) with RichIterator
+    val iter = new RichStringIter
     iter foreach println
   }
 }
 ```
```

Comment:
Or just paste into REPL.

I'd suggest losing the `main` method, which also uses procedure syntax here, and use `App` when called for. I'm a bigger fan of just pasting snippets, using the REPL as intended to show values and avoiding println everywhere. My least favorite style is a custom entry point like `runTest` as shown. I'm not above `MyApp main null` in REPL.